LibWeb: Speed up CSS ValueID parsing by factor 20

BenWiederhake opened the pull request:
The old approach was basically a linear scan, which is slower than a hash map for the currently 303 elements. In the course of doing so, this also uncovers and fixes a bug in AK::CaseInsensitiveStringViewTraits.

nico commented:
Is there an absolute speedup as well? I can't imagine a linear scan over 300 elements taking any actual time.

BenWiederhake commented:
Oh! I forgot to paste the time measurements. Yes, there is an absolute speedup, as measured by the new benchmark in TestCSSIDSpeed.cpp. Thankfully I still have the results in my console buffer:

Before: Completed benchmark 'value_id_from_string' in 3238ms
After: Completed benchmark 'value_id_from_string' in 193ms

That's also where I got the number "20" from, because 3298 / 193 is 17 (which I eyeballed as 20).

AtkinsSJ commented:
> Do you know how often `value_id_from_string` is called in practice?

Quite a lot. Every single CSS identifier being parsed goes through that, and there tend to be a lot of them.

(Also, parsing string, image and rect values calls that because the value-parsing code just tries one type after another until one works. :sweat_smile:)

BenWiederhake commented:
TestHTMLTokenizer used to be a dead test, it always existed but was never built. I neglected to run it, apparently it's broken, and that's why CI fails.

AtkinsSJ commented:
> TestHTMLTokenizer used to be a dead test, it always existed but was never built. I neglected to run it, apparently it's broken, and that's why CI fails.

Enjoy your free yak! Enabling the tests probably wants to go in its own commit too.

@ADKaster does the cmake stuff look correct to you?

davidot commented on Tests/LibWeb/TestCSSIDSpeed.cpp:
```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright (c) 2021, Max Wipfli <max.wipfli@serenityos.org>
```
Is this copyright correct?

BenWiederhake replied:
You're right! Fixed it in the recent push.

davidot commented on Meta/Lagom/CMakeLists.txt:
```diff
@@ -657,6 +657,15 @@ if (BUILD_LAGOM)
         endforeach()
 
+        # Web
+        if (ENABLE_LAGOM_LIBWEB)
+            file(GLOB LIBWEB_TEST_SOURCES CONFIGURE_DEPENDS "../../Tests/LibWeb/*.cpp")
+            foreach(source ${LIBWEB_TEST_SOURCES})
+                lagom_test(${source} LIBS LibWeb
+                    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/../../Tests/LibWeb)
+            endforeach()
+        endif()
```
I think we do prefer not globbing (like the js tests below) but we also do it above here so :shrug:

BenWiederhake replied:
I decided against removing globbing, because we do that everywhere and isn't it nicer when one doesn't have to manually add a new Test file? ^^

davidot reviewed (changes requested):
One small comment and feel free to ignore the globbing one

BenWiederhake commented:
Changes since first push:
- Rebased for convenience
- The failing test in TestHTMLTokenizer depends on a file in a hard-coded location. I took the same approach as TestBrotli, and wrapped the path in `#ifdef AK_OS_SERENITY`, and cmake now copies the file into the build dir, which tests successfully.
- Fixed the Copyright. Thanks @davidot!
- I split the commits as suggested, this also makes it easier to verify my claim about the speedup. Thanks @davidot!
- I decided against removing globbing, because we do that everywhere and isn't it nicer when one doesn't have to manually add a new Test file? ^^

ADKaster reviewed (approved):
I am ok with these changes as-is but a few comments:

>  decided against removing globbing, because we do that everywhere and isn't it nicer when one doesn't have to manually add a new Test file? ^^

This is very much not the right position to take. GLOBs are bad because they force the generator (ninja, in our case normally) to recompute changed files based on timestamps rather than relying on update CMakeLists --> update generator files. It creates back references in the build system that slow down the build. 

We should be looking to eliminate ALL globs from our build system. I've taken us most of the way there, but we still need to come up with a way to include_directories(Tests) and have it Just Work :tm: for lagom. Probably needs a `serenity_test` utility we can override in Lagom. 

GLOBs cause actual problems when we remove or move files, as shown by recent issues with the AK/FlyString.cpp move.

BenWiederhake commented:
Oh! Thanks, I didn't know it was causing such issues. I'll see whether I can help with it :)
